File: admin_platform/cpp/src/handlers/user_handler.cpp
/**
 * TigerWallet Admin Platform - C++ User Handler Implementation
 * High-performance user management handler
 */

#include "../include/handlers/user_handler.h"
#include "../include/database.h"
#include "../include/redis_client.h"
#include <iostream>
#include <sstream>
#include <regex>

namespace tiger {
namespace handlers {

UserHandler::UserHandler(std::shared_ptr<Database> db, std::shared_ptr<RedisClient> redis)
    : db_(db), redis_(redis) {}

nlohmann::json UserHandler::get_users(const nlohmann::json& request) {
    int page = request.value("page", 1);
    int limit = request.value("limit", 20);
    std::string status = request.value("status", "");
    std::string search = request.value("search", "");
    
    std::stringstream query;
    query << "SELECT id, user_id, username, email, phone, status, tier, kyc_status, "
          << "kyc_level, is_email_verified, is_phone_verified, white_label_id, "
          << "referral_code, created_at, last_login FROM users WHERE 1=1";
    
    if (!status.empty()) {
        query << " AND status = '" << db_->escape_string(status) << "'";
    }
    
    if (!search.empty()) {
        query << " AND (username ILIKE '%" << db_->escape_string(search) << "%' "
              << "OR email ILIKE '%" << db_->escape_string(search) << "%')";
    }
    
    query << " ORDER BY created_at DESC";
    query << " LIMIT " << limit << " OFFSET " << ((page - 1) * limit);
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to fetch users"},
            {"success", false}
        });
    }
    
    nlohmann::json users = nlohmann::json::array();
    
    for (int i = 0; i < result->rows(); ++i) {
        users.push_back(nlohmann::json::object({
            {"id", result->get_value(i, "id")},
            {"user_id", result->get_value(i, "user_id")},
            {"username", result->get_value(i, "username")},
            {"email", result->get_value(i, "email")},
            {"phone", result->get_value(i, "phone")},
            {"status", result->get_value(i, "status")},
            {"tier", std::stoi(result->get_value(i, "tier"))},
            {"kyc_status", result->get_value(i, "kyc_status")},
            {"kyc_level", std::stoi(result->get_value(i, "kyc_level"))},
            {"is_email_verified", result->get_value(i, "is_email_verified") == "true"},
            {"is_phone_verified", result->get_value(i, "is_phone_verified") == "true"},
            {"white_label_id", result->get_value(i, "white_label_id")},
            {"referral_code", result->get_value(i, "referral_code")},
            {"created_at", result->get_value(i, "created_at")},
            {"last_login", result->get_value(i, "last_login")}
        }));
    }
    
    // Get total count
    std::stringstream count_query;
    count_query << "SELECT COUNT(*) as total FROM users WHERE 1=1";
    if (!status.empty()) {
        count_query << " AND status = '" << db_->escape_string(status) << "'";
    }
    auto count_result = db_->execute(count_query.str());
    int total = count_result && count_result->rows() > 0 ? 
        std::stoi(count_result->get_value(0, "total")) : 0;
    
    return nlohmann::json::object({
        {"success", true},
        {"data", users},
        {"meta", nlohmann::json::object({
            {"page", page},
            {"limit", limit},
            {"total", total},
            {"total_pages", (total + limit - 1) / limit}
        })}
    });
}

nlohmann::json UserHandler::get_user(const std::string& user_id) {
    std::stringstream query;
    query << "SELECT id, user_id, username, email, phone, status, tier, kyc_status, "
          << "kyc_level, is_email_verified, is_phone_verified, white_label_id, "
          << "referral_code, created_at, last_login FROM users "
          << "WHERE id = '" << db_->escape_string(user_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result || result->rows() == 0) {
        return nlohmann::json::object({
            {"error", "User not found"},
            {"success", false}
        });
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"data", nlohmann::json::object({
            {"id", result->get_value(0, "id")},
            {"user_id", result->get_value(0, "user_id")},
            {"username", result->get_value(0, "username")},
            {"email", result->get_value(0, "email")},
            {"phone", result->get_value(0, "phone")},
            {"status", result->get_value(0, "status")},
            {"tier", std::stoi(result->get_value(0, "tier"))},
            {"kyc_status", result->get_value(0, "kyc_status")},
            {"kyc_level", std::stoi(result->get_value(0, "kyc_level"))},
            {"is_email_verified", result->get_value(0, "is_email_verified") == "true"},
            {"is_phone_verified", result->get_value(0, "is_phone_verified") == "true"},
            {"white_label_id", result->get_value(0, "white_label_id")},
            {"referral_code", result->get_value(0, "referral_code")},
            {"created_at", result->get_value(0, "created_at")},
            {"last_login", result->get_value(0, "last_login")}
        })}
    });
}

nlohmann::json UserHandler::update_user(const std::string& user_id, const nlohmann::json& data) {
    std::vector<std::string> updates;
    
    if (data.contains("username")) {
        updates.push_back("username = '" + db_->escape_string(data["username"].get<std::string>()) + "'");
    }
    if (data.contains("email")) {
        updates.push_back("email = '" + db_->escape_string(data["email"].get<std::string>()) + "'");
    }
    if (data.contains("phone")) {
        updates.push_back("phone = '" + db_->escape_string(data["phone"].get<std::string>()) + "'");
    }
    if (data.contains("status")) {
        updates.push_back("status = '" + db_->escape_string(data["status"].get<std::string>()) + "'");
    }
    if (data.contains("kyc_status")) {
        updates.push_back("kyc_status = '" + db_->escape_string(data["kyc_status"].get<std::string>()) + "'");
    }
    
    if (updates.empty()) {
        return nlohmann::json::object({
            {"error", "No fields to update"},
            {"success", false}
        });
    }
    
    std::stringstream query;
    query << "UPDATE users SET " << updates[0];
    for (size_t i = 1; i < updates.size(); ++i) {
        query << ", " << updates[i];
    }
    query << ", updated_at = NOW() WHERE id = '" << db_->escape_string(user_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to update user"},
            {"success", false}
        });
    }
    
    // Invalidate cache
    if (redis_) {
        redis_->del("user:" + user_id);
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "User updated successfully"}
    });
}

nlohmann::json UserHandler::suspend_user(const std::string& user_id, const std::string& reason) {
    std::stringstream query;
    query << "UPDATE users SET status = 'suspended', "
          << "suspend_reason = '" << db_->escape_string(reason) << "', "
          << "suspended_at = NOW() WHERE id = '" << db_->escape_string(user_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to suspend user"},
            {"success", false}
        });
    }
    
    // Invalidate all user sessions
    if (redis_) {
        redis_->del("session:user:" + user_id);
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "User suspended successfully"}
    });
}

nlohmann::json UserHandler::ban_user(const std::string& user_id, const std::string& reason) {
    std::stringstream query;
    query << "UPDATE users SET status = 'banned', "
          << "ban_reason = '" << db_->escape_string(reason) << "', "
          << "banned_at = NOW() WHERE id = '" << db_->escape_string(user_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to ban user"},
            {"success", false}
        });
    }
    
    // Invalidate all user sessions and tokens
    if (redis_) {
        redis_->del("session:user:" + user_id);
        redis_->del("token:user:" + user_id);
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "User banned successfully"}
    });
}

nlohmann::json UserHandler::activate_user(const std::string& user_id) {
    std::stringstream query;
    query << "UPDATE users SET status = 'active', "
          << "suspend_reason = NULL, suspended_at = NULL, "
          << "ban_reason = NULL, banned_at = NULL "
          << "WHERE id = '" << db_->escape_string(user_id) << "'";
    
    auto result = db_->execute(query.str());
    
    if (!result) {
        return nlohmann::json::object({
            {"error", "Failed to activate user"},
            {"success", false}
        });
    }
    
    return nlohmann::json::object({
        {"success", true},
        {"message", "User activated successfully"}
    });
}

} // namespace handlers
} // namespace tiger

File: admin/go/internal/handlers/user_handler.go
package handlers

import (
	"net/http"
	"strconv"
	"time"

	"github.com/tigerwallet/admin/internal/models"
	"github.com/tigerwallet/admin/pkg/database"

	"github.com/gin-gonic/gin"
)

// UserHandler handles user-related requests
type UserHandler struct {
	db *database.PostgresDB
}

// NewUserHandler creates a new user handler
func NewUserHandler(db *database.PostgresDB) *UserHandler {
	return &UserHandler{db: db}
}

// ListUsers lists all users with pagination and filters
func (h *UserHandler) ListUsers(c *gin.Context) {
	page := c.DefaultQuery("page", "1")
	pageSize := c.DefaultQuery("page_size", "20")
	search := c.Query("search")
	status := c.Query("status")
	kycStatus := c.Query("kyc_status")
	whiteLabelID := c.Query("white_label_id")

	pageInt, _ := strconv.Atoi(page)
	pageSizeInt, _ := strconv.Atoi(pageSize)

	var users []models.User
	var total int64

	query := h.db.Model(&models.User{})

	// Apply filters
	if search != "" {
		query = query.Where("email ILIKE ? OR username ILIKE ? OR wallet_address ILIKE ?",
			"%"+search+"%", "%"+search+"%", "%"+search+"%")
	}
	if status != "" {
		query = query.Where("status = ?", status)
	}
	if kycStatus != "" {
		query = query.Where("kyc_status = ?", kycStatus)
	}
	if whiteLabelID != "" {
		query = query.Where("white_label_id = ?", whiteLabelID)
	}

	// Count total
	query.Count(&total)

	// Paginate
	offset := (pageInt - 1) * pageSizeInt
	query = query.Offset(offset).Limit(pageSizeInt).Order("created_at DESC")

	if err := query.Find(&users).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch users"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":        users,
		"total":       total,
		"page":        pageInt,
		"page_size":   pageSizeInt,
		"total_pages": (total + int64(pageSizeInt) - 1) / int64(pageSizeInt),
	})
}

// GetUser gets a user by ID
func (h *UserHandler) GetUser(c *gin.Context) {
	userID := c.Param("id")

	var user models.User
	if err := h.db.Preload("KYCApplications").First(&user, userID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	c.JSON(http.StatusOK, user)
}

// UpdateUser updates a user
func (h *UserHandler) UpdateUser(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	userID := c.Param("id")

	var req struct {
		Status   string `json:"status"`
		KYCLevel int    `json:"kyc_level"`
		Tags     string `json:"tags"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	var user models.User
	if err := h.db.First(&user, userID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	updates := map[string]interface{}{}
	if req.Status != "" {
		updates["status"] = req.Status
	}
	if req.KYCLevel > 0 {
		updates["kyc_level"] = req.KYCLevel
	}

	if err := h.db.Model(&user).Updates(updates).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to update user"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "update_user", "user", userID, "User updated", c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, user)
}

// DeleteUser soft deletes a user
func (h *UserHandler) DeleteUser(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	userID := c.Param("id")

	if err := h.db.Model(&models.User{}).Where("id = ?", userID).Update("status", "deleted").Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete user"})
		return
	}

	// Log activity
	logAdminActivity(h.db, adminID, "delete_user", "user", userID, "User deleted", c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "User deleted successfully"})
}

// VerifyKYC verifies a user's KYC
func (h *UserHandler) VerifyKYC(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	userID := c.Param("id")

	var req struct {
		Level int    `json:"level" binding:"required"`
		Notes string `json:"notes"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	var user models.User
	if err := h.db.First(&user, userID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
		return
	}

	now := time.Now()
	if err := h.db.Model(&user).Updates(map[string]interface{}{
		"kyc_status":      "level" + strconv.Itoa(req.Level),
		"kyc_level":       req.Level,
		"kyc_verified_at": now,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to verify KYC"})
		return
	}

	// Create KYC application record
	kycApp := models.KYCApplication{
		UserID:      user.ID,
		Level:       req.Level,
		Status:      "approved",
		SubmittedAt: now,
		ReviewedAt:  &now,
		ReviewedBy:  &adminID,
		Notes:       req.Notes,
		IPAddress:   c.ClientIP(),
	}
	h.db.Create(&kycApp)

	// Log activity
	logAdminActivity(h.db, adminID, "verify_kyc", "user", userID, "KYC verified at level "+strconv.Itoa(req.Level), c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "KYC verified successfully"})
}

// DashboardHandler handles dashboard requests
type DashboardHandler struct {
	db *database.PostgresDB
}

// NewDashboardHandler creates a new dashboard handler
func NewDashboardHandler(db *database.PostgresDB) *DashboardHandler {
	return &DashboardHandler{db: db}
}

// GetDashboard gets dashboard statistics
func (h *DashboardHandler) GetDashboard(c *gin.Context) {
	var stats DashboardStats

	// Total users
	h.db.Model(&models.User{}).Count(&stats.TotalUsers)

	// Active users
	h.db.Model(&models.User{}).Where("status = ?", "active").Count(&stats.ActiveUsers)

	// Pending KYC
	h.db.Model(&models.User{}).Where("kyc_status = ?", "pending").Count(&stats.PendingKYC)

	// Today's new users
	today := time.Now().Truncate(24 * time.Hour)
	h.db.Model(&models.User{}).Where("created_at >= ?", today).Count(&stats.NewUsersToday)

	// Today's transactions
	h.db.Model(&models.Transaction{}).Where("created_at >= ?", today).Count(&stats.TodayTransactions)

	stats.SystemHealth = "99.9%"

	c.JSON(http.StatusOK, stats)
}

// TransactionHandler handles transaction requests
type TransactionHandler struct {
	db *database.PostgresDB
}

// NewTransactionHandler creates a new transaction handler
func NewTransactionHandler(db *database.PostgresDB) *TransactionHandler {
	return &TransactionHandler{db: db}
}

// ListTransactions lists all transactions
func (h *TransactionHandler) ListTransactions(c *gin.Context) {
	page := c.DefaultQuery("page", "1")
	pageSize := c.DefaultQuery("page_size", "20")
	txType := c.Query("type")
	status := c.Query("status")
	chain := c.Query("chain")

	pageInt, _ := strconv.Atoi(page)
	pageSizeInt, _ := strconv.Atoi(pageSize)

	var transactions []models.Transaction
	var total int64

	query := h.db.Model(&models.Transaction{})

	if txType != "" {
		query = query.Where("type = ?", txType)
	}
	if status != "" {
		query = query.Where("status = ?", status)
	}
	if chain != "" {
		query = query.Where("chain = ?", chain)
	}

	query.Count(&total)

	offset := (pageInt - 1) * pageSizeInt
	query = query.Offset(offset).Limit(pageSizeInt).Order("created_at DESC").Preload("User")

	if err := query.Find(&transactions).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch transactions"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"data":        transactions,
		"total":       total,
		"page":        pageInt,
		"page_size":   pageSizeInt,
		"total_pages": (total + int64(pageSizeInt) - 1) / int64(pageSizeInt),
	})
}

// GetTransaction gets a transaction by ID
func (h *TransactionHandler) GetTransaction(c *gin.Context) {
	txID := c.Param("id")

	var tx models.Transaction
	if err := h.db.Preload("User").First(&tx, txID).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Transaction not found"})
		return
	}

	c.JSON(http.StatusOK, tx)
}

// FlagTransaction flags a transaction
func (h *TransactionHandler) FlagTransaction(c *gin.Context) {
	adminID := c.GetUint("admin_id")
	txID := c.Param("id")

	var req struct {
		Reason string `json:"reason" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	if err := h.db.Model(&models.Transaction{}).Where("id = ?", txID).Updates(map[string]interface{}{
		"flagged":     true,
		"flag_reason": req.Reason,
	}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to flag transaction"})
		return
	}

	logAdminActivity(h.db, adminID, "flag_transaction", "transaction", txID, "Transaction flagged: "+req.Reason, c.ClientIP(), c.Request.UserAgent())

	c.JSON(http.StatusOK, gin.H{"message": "Transaction flagged"})
}

// Helper function to log admin activity
func logAdminActivity(db *database.PostgresDB, adminID uint, action, resource, resourceID, details, ip, userAgent string) {
	activity := models.AdminActivity{
		AdminID:    adminID,
		Action:     action,
		Resource:   resource,
		ResourceID: resourceID,
		IPAddress:  ip,
		UserAgent:  userAgent,
		Status:     "success",
	}
	db.Create(&activity)
}
